### ISSUE TITLE
Add exhaustive mock-test coverage of ``CheckTransaction``

### ISSUE THREAD

**Issue by nathan-at-least** [zcash/zcash#863]
After #862, #858, and #860 are all complete, this ticket is closed when ``CheckTransaction`` unit tests:

a. exhaustively cover every branch in ``CheckTransaction``,
b. the fact of a. is observable from reports automatically generated by buildbot in the PR for this ticket,
c. the unit test does not rely on I/O, threads, or globals, and has no side-effects.

**Comment by defuse** [comment]
WIP of this in #1051 (Hopefully I didn't duplicate any work @daira had already done).

The next challenge for this is to figure out how to mock CPourTx's Verify() method. The way I'm familiar with in gmock requires making a subclass, and so we could put an instance of the mock into `vpour`.

**Comment by defuse** [comment]
It looks like the options for mocking out the pour verification in CheckTransaction are:

1. Change CMutableTransaction to a template, and parameterize what `vpour` is a vector of. In production it will be a `CMutableTransaction<CPourTx>` and in testing it will be a `CMutableTransaction<MockCPourTx>`.
2. Change `vpour` to be not a vector of `CPourTx` but instead a vector of pointers to `CPourTx` (so that polymorphism works).
3. (Sean's idea): Split the "hard stuff" like pour verification out of CheckTransaction, so that CheckTransaction doesn't do the pour verification, and we can mock test all the cases inside the simpler CheckTransaction, then have a CheckTransactionAndProof that calls CheckTransaction and then checks the proof.

None of these seem ideal to me, any other ideas?

**Comment by defuse** [comment]
Actually I like (3) because it makes it easy to see that only the tests are calling CheckTransactionWithoutProofVerification, instead of having to set a flag inside the state that gets passed in (which could accidentally be set in production).

**Comment by defuse** [comment]
In our pairing me and Sean went with option 3. I like the results, visible in PR #1051.

**Comment by defuse** [comment]
This is done, just blocking on merging of #1051.
